<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>4. The assemble module &mdash; bgrrl 0.6 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="bgrrl 0.6 documentation" href="../index.html" />
    <link rel="next" title="5. The annotate module" href="annotate.html" />
    <link rel="prev" title="3. The survey module" href="survey.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-assemble-module">
<h1>4. The assemble module<a class="headerlink" href="#the-assemble-module" title="Permalink to this headline">¶</a></h1>
<img alt="../_images/bgrrl_workflow.png" class="align-center" src="../_images/bgrrl_workflow.png" />
<p>The assemble module performs automatically optimized assemblies of the provided library data.</p>
<div class="section" id="sample-sheet-preparation">
<h2>4.1. Sample sheet preparation<a class="headerlink" href="#sample-sheet-preparation" title="Permalink to this headline">¶</a></h2>
<p>If the user has previously run the <code class="docutils literal"><span class="pre">survey</span></code> module, the resulting samplesheet <code class="docutils literal"><span class="pre">samplesheet.qc_pass.tsv</span></code> can be used directly
to drive the assemble module.</p>
<p>Otherwise, the user should prepare a comma-separated samplesheet following the column order below.</p>
<ol class="arabic simple">
<li>Sample ID</li>
<li>Sample Name (can be the same as Sample ID; the intention is to allow a more understandable sample reference in the future)</li>
<li>Full path to R1 file</li>
<li>Full path to R2 file</li>
<li>LEAVE EMPTY (intended use: Full path to single-end file)</li>
<li>LEAVE EMPTY (intended use: NCBI taxonomy id)</li>
<li>LEAVE EMPTY (intended use: Taxonomy name)</li>
<li>LEAVE EMPTY (intended use: FastQC report for raw R1)</li>
<li>LEAVE EMPTY (intended use: FastQC report for raw R2)</li>
<li>LEAVE EMPTY (intended use: FastQC report for raw single-end file)</li>
<li>Full path to preprocessed, non-normalized R1 file</li>
<li>Full path to preprocessed, non-normalized R2 file</li>
<li>LEAVE EMPTY (intended use: Full path to preprocessed, non-normalized single-end file)</li>
<li>Full path to preprocessed, normalized R1 file</li>
<li>Full path to preprocessed, normalized R2 file</li>
<li>LEAVE EMPTY (intended use: Full path to preprocessed, normalized single-end file)</li>
</ol>
</div>
<div class="section" id="command-line-arguments">
<h2>4.2. Command line arguments<a class="headerlink" href="#command-line-arguments" title="Permalink to this headline">¶</a></h2>
<p>The command <code class="docutils literal"><span class="pre">bgrrl</span> <span class="pre">-h</span></code> or <code class="docutils literal"><span class="pre">bgrrl</span> <span class="pre">&lt;stage&gt;</span> <span class="pre">-h</span></code> (or <code class="docutils literal"><span class="pre">--help</span></code> instead of <code class="docutils literal"><span class="pre">-h</span></code>) will display a list of command line options.</p>
<p><strong>Remember each bgrr| run requires at the very least the following three command line parameters:</strong></p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">input_sheet</span></code></li>
<li><code class="docutils literal"><span class="pre">--config</span></code></li>
<li><code class="docutils literal"><span class="pre">--hpc_config</span></code></li>
</ul>
<div class="section" id="assemble-options">
<h3>4.2.1. assemble options:<a class="headerlink" href="#assemble-options" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first"><code class="docutils literal"><span class="pre">--assembler</span> <span class="pre">{unicycler,velvet}</span></code></p>
<blockquote>
<div><p>Assembly software to use for genome assembly. [unicycler]</p>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--contig-minlen</span> <span class="pre">CONTIG_MINLEN</span></code></p>
<blockquote>
<div><p>Minimum length [bp] of contigs retained in filtering step [0].</p>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--no-normalization</span></code></p>
<p>Use non-normalized reads in assemble module [False]</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--run-annotation</span></code></p>
<dl class="docutils">
<dt>Run annotation on assembly. If set, de novo annotation</dt>
<dd><p class="first last">with prokka will be run. Additionally, you may enable
annotation transfer by specifying a path to a
reference annotation with <code class="docutils literal"><span class="pre">--ratt-reference</span></code>. [False]</p>
</dd>
</dl>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--custom-prokka-proteins</span> <span class="pre">CUSTOM_PROKKA_PROTEINS</span></code></p>
<blockquote>
<div><p>If you have a custom protein database that you would
like prokka to use (<code class="docutils literal"><span class="pre">prokka</span></code>&#8216;s <code class="docutils literal"><span class="pre">--proteins</span></code> option), then specify
the path to it here. [n/a]</p>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--ratt-reference</span> <span class="pre">RATT_REFERENCE</span></code></p>
<blockquote>
<div><p>Path to reference data for ratt annotation transfer</p>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--is-final-step</span></code></p>
<blockquote>
<div><dl class="docutils">
<dt>If set, analysis packaging will take place after the</dt>
<dd><p class="first last">assembly stage. Otherwise, assume that an annotation
stage will follow, which will then take care of
analysis packaging. [False]</p>
</dd>
</dl>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--no-packaging</span></code></p>
<blockquote>
<div><p>Disable automatic packaging. [False]</p>
</div></blockquote>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--prokka-package-style</span> <span class="pre">{by_sample,all_in_one}</span></code></p>
<blockquote>
<div><p>Should the prokka annotation be packaged into one
directory per sample (by_sample) or into one single
directory (all_in_one)? [by_sample]</p>
</div></blockquote>
</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">4. The assemble module</a><ul>
<li><a class="reference internal" href="#sample-sheet-preparation">4.1. Sample sheet preparation</a></li>
<li><a class="reference internal" href="#command-line-arguments">4.2. Command line arguments</a><ul>
<li><a class="reference internal" href="#assemble-options">4.2.1. assemble options:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="survey.html" title="previous chapter">3. The survey module</a></li>
      <li>Next: <a href="annotate.html" title="next chapter">5. The annotate module</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/Usage/assemble.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Christian Schudoma.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.8</a>
      
      |
      <a href="../_sources/Usage/assemble.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>